src/index_with_decompressor.ts

```diff
@@ -0,0 +1,120 @@
+import { Chess } from '../cjsmin/src/chess';
+import { gameChunks } from './fileReader';
+import { KDRatioMetric, MateAndAssistMetric, KillStreakMetric } from './metrics/captures';
+import { MoveDistanceMetric } from './metrics/distances';
+import { MetadataMetric } from './metrics/misc';
+import {
+  GameWithMostMovesMetric,
+  PieceLevelMoveInfoMetric,
+  MiscMoveFactMetric,
+} from './metrics/moves';
+import { PromotionMetric } from './metrics/promotions';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as async from 'async';
+
+/**
+ *
+ * @param path
+ * @returns
+ */
+export async function main(path: string) {
+  console.time('Total Execution Time');
+  await gameIterator(path);
+  console.timeEnd('Total Execution Time');
+  return results;
+}
+
+let results = {
+  'Number of games analyzed': 0,
+}
+
+/**
+ * Metric functions will ingest a single game at a time
+ * @param metricFunctions
+ */
+async function gameIterator(path) {
+  const cjsmin = new Chess();
+
+  const gamesGenerator = gameChunks(path);
+  const kdRatioMetric = new KDRatioMetric();
+  const killStreakMetric = new KillStreakMetric();
+  const mateAndAssistMetric = new MateAndAssistMetric();
+  const promotionMetric = new PromotionMetric();
+  const moveDistanceMetric = new MoveDistanceMetric();
+  const gameWithMostMovesMetric = new GameWithMostMovesMetric();
+  const pieceLevelMoveInfoMetric = new PieceLevelMoveInfoMetric();
+  const metadataMetric = new MetadataMetric(cjsmin);
+  const miscMoveFactMetric = new MiscMoveFactMetric();
+  const metrics = [
+    metadataMetric,
+    kdRatioMetric,
+    killStreakMetric,
+    mateAndAssistMetric,
+    promotionMetric,
+    moveDistanceMetric,
+    gameWithMostMovesMetric,
+    pieceLevelMoveInfoMetric,
+    miscMoveFactMetric,
+  ];
+
+  let gameCounter = 0;
+  for await (const { moves, metadata } of gamesGenerator) {
+    gameCounter++;
+    if (gameCounter % 400 == 0) {
+      console.log('number of games ingested: ', gameCounter);
+    }
+
+    for (const metric of metrics) {
+      // with array creation
+      const historyGenerator = cjsmin.historyGeneratorArr(moves);
+      metric.processGame(Array.from(historyGenerator), metadata);
+    }
+  }
+  results['Number of games analyzed'] = gameCounter;
+  let metricCallsCount = 0;
+  for (const metric of metrics) {
+    metricCallsCount++;
+    results[metric.constructor.name] = metric.aggregate()
+  }
+}
+
+// Create a write to result.json queue with a concurrency of 1
```

Comment:
I know you were having issues with concurrency when running multiple processes - this queue is going to live inside of each process, you'd need a single process with the queue for it to function as expected.

My recommendation would be to have each process write to an intermediate aggregate file then combine those - probably the easiest to set up since we wouldn't have ot . You might have to track some additional data so computations could be done correctly (weighted averages), or you should just compute totals in the intermediate step

Reply:
I think this makes sense 100%. I think you might not have to track additional data if you make as many intermediate aggregate files (e.g. "results1.json, results2.json... results1500.json")